_posts/2021-12-28-trino-2021-a-year-of-growth.md

```diff
@@ -0,0 +1,347 @@
+
+## Data Sources
+
+### BigQuery connector
+
+* Added support for case insensitive name matching for BigQuery views.
+* Support reading `bignumeric` type whose precision is less than or equal to 
+38.
+* Add support for `CREATE SCHEMA` and `DROP SCHEMA` statements.
+* Improved support for BigQuery datetime and timestamp types.
+* Fixed failures for queries accessing `information_schema.columns` when 
+`case-insensitive-name-matching` is disabled.
+
+### Cassandra connector
+
+* Mapped Cassandra `uuid` type to Trino `uuid`.
+* Added support for Cassandra `tuple` type.
+* Changed minimum number of speculative executions from 2 to 1. 
+
+### Clickhouse connector
+
+* Added [ClickHouse connector]({{site.url}}/docs/current/connector/clickhouse.html).
+* Improved performance of aggregation queries by computing aggregations within 
+ClickHouse. Currently, the following aggregate functions are eligible for
+pushdown: `count`, `min`, `max`, `sum` and `avg`.
+* Added support for dropping columns.
+* Map ClickHouse `UUID` columns as `UUID` type in Trino instead of `VARCHAR`.
+
+### HDFS, S3, Azure and cloud object storage systems
+
+A core use case of Trino uses the Hive and Iceberg connectors to connect to
+a data lake. These connectors differ from most as Trino is the sole query engine
+as opposed to the client calling another system. Here are some changes that
+for these connectors:
+
+* Fixed queries on empty Iceberg tables without a snapshot ID that were created by Spark.
+* Updated to Iceberg 0.11.0 behavior for transforms of dates and timestamps before 1970.
+* Added procedure `system.flush_metadata_cache()` to flush metadata caches.
+* Avoid generating splits for empty files.
+* Sped up Iceberg query performance when dynamic filtering can be leveraged.
+* Increased Iceberg performance when reading timestamps from Parquet files.
+* Improved Iceberg performance for queries on nested data through dereference pushdown.
+* Added support for `INSERT OVERWRITE` operations on S3-backed tables.
+* Made the Iceberg `uuid` type available.
+* Trino views made available in Iceberg.
+
+
+### Elasticsearch connector
+
+* Added support for reading fields as `json` values.
+* Fixed failure when documents contain fields of unsupported types.
+* Added support for `scaled_float` type.
+* Added support for assuming an IAM role.
+* Fixed potential incorrect results when queries contain an `IS NULL` predicate.
+* Added retry requests with backoff when Elasticsearch is overloaded.
+
+### Kafka connector
+
+* Added support for TLS security protocol.
+* Fixed querying of schema registry tables with references in their schema.
+* Fixed listing of schema registry tables having ambiguous subject name in lower case.
+
+### MemSQL connector
+
+* Allowed limiting the size of the metadata cache.
+* Added partial support for `DELETE` statement where predicate can be fully pushed 
+down to the remote datasource.
```

Comment:
These apply to all JDBC connectors.

Reply:
Also handling of older dates (Julian to Gregorian switch) was improved in many JDBC connectors to be made consistent with Trino's semantics.

Are we trying to include everything from past release notes? How do we decide what to keep and what to drop?
e.g. JDBC connectors also gained the `flush_metadata_cache` procedure.

Reply:
I had a list that I thought was important...I didn't really get a list from others so I am just guessing here and hoping for suggestions or feedback from others. Manfred and I are going to try and get ahead of this next year and try to push people to make contributions to this list as releases come out.

Reply:
We could take out the MemSQL connector section if most of what was added to it are general JDBC features.

Reply:
There are some general JDBC related improvements which I believe you can pull out to a separate section:

1. statistical aggregate function pushdown - https://trino.io/docs/current/release/release-352.html#mysql-connector (and other releases)
2. topn pushdown and join pushdown - https://trino.io/docs/current/release/release-353.html
3. Improved planning times by reducing number of connections opened - https://trino.io/docs/current/release/release-353.html
4. Improved performance by improving metadata caching hit rate -https://trino.io/docs/current/release/release-356.html
5. Rule based identifier mapping support - https://trino.io/docs/current/release/release-357.html
6. DELETE, non-transactional inserts and write-batch-size - https://trino.io/docs/current/release/release-360.html
7. Metadata cache max size - https://trino.io/docs/current/release/release-361.html
8. TRUNCATE - https://trino.io/docs/current/release/release-365.html
9. Improved handling of Gregorian - Julian switch for `date` type - https://trino.io/docs/current/release/release-366.html
10. Done across multiple releases - ensured correctness when pushing down predicates/topN to remote system that is case-insensisitive or sorts differently than Trino.

Reply:
I had one before. I do like the idea of it but @mosabua was concerned about it. Or being clear of which connectors are jdbc and if it would be too verbose to list it out. WDYT?

I think we should add these and just list it out.

Reply:
We can always word it like `Some improvements were also made which apply to some/most JDBC connectors:`

Reply:
I'm fine either way - whatever is less confusing or easy to do.